pkg/operators/filter/filter.go

```diff
@@ -0,0 +1,389 @@
+	"fmt"
+	"strconv"
+	"strings"
+
+	"golang.org/x/exp/constraints"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+type comparisonType int
+
+const (
+	ParamFilter = "filter"
+)
+
+const (
+	comparisonTypeUnknown comparisonType = iota
+	comparisonTypeMatch
+	comparisonTypeRegex
+	comparisonTypeLt
+	comparisonTypeLte
+	comparisonTypeGt
+	comparisonTypeGte
+)
+
+type filterOperator struct{}
+
+func (f *filterOperator) Name() string {
+	return "filter"
+}
+
+func (f *filterOperator) Init(params *params.Params) error {
+	return nil
+}
+
+func (f *filterOperator) GlobalParams() api.Params {
+	return nil
+}
+
+func (f *filterOperator) InstanceParams() api.Params {
+	return api.Params{&api.Param{
+		Key: ParamFilter,
+		Description: `Filter rules
+  A filter can match any column using the following syntax
+    columnName==value     - matches, if the content of columnName equals exactly value
+    columnName!=value     - matches, if the content of columnName does not equal exactly value
+    columnName>=value     - matches, if the content of columnName is greater than or equal to the value
+    columnName>value      - matches, if the content of columnName is greater than the value
+    columnName<=value     - matches, if the content of columnName is less than or equal to the value
+    columnName<value      - matches, if the content of columnName is less than the value
+    columnName~value      - matches, if the content of columnName matches the regular expression 'value'
+                 see [https://github.com/google/re2/wiki/Syntax] for more information on the syntax
+        `,
+		Alias: "F",
+	}}
+}
+
+func (f *filterOperator) InstantiateDataOperator(gadgetCtx operators.GadgetContext, instanceParamValues api.ParamValues) (operators.DataOperatorInstance, error) {
+	filterCfg, _ := instanceParamValues[ParamFilter]
+
+	fop := &filterOperatorInstance{
+		ffns: map[datasource.DataSource][]func(datasource.DataSource, datasource.Data) bool{},
+	}
+
+	filters := strings.Split(filterCfg, ",")
+	for _, filter := range filters {
+		if filter == "" {
+			continue
+		}
+		err := fop.addFilter(gadgetCtx, filter)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return fop, nil
+}
+
+func (f *filterOperator) Priority() int {
+	return 49000
+}
+
+type filterOperatorInstance struct {
+	ffns map[datasource.DataSource][]func(datasource.DataSource, datasource.Data) bool
+}
+
+func (f *filterOperatorInstance) Name() string {
+	return "filter"
+}
+
+func (f *filterOperatorInstance) PreStart(gadgetCtx operators.GadgetContext) error {
+	for ds, funcs := range f.ffns {
+		funcs := funcs
+		ds.Subscribe(func(ds datasource.DataSource, data datasource.Data) error {
+			for _, fn := range funcs {
+				if !fn(ds, data) {
+					return fmt.Errorf("skip")
```

Comment:
What's your idea here? This approach wouldn't work for arrays. I was wondering how we could allow subscribers to "discard" data. I'm thinking to what we do for enriching containers where enrichers can return false if they want the container to be discarded. Another way is to expose a method to mark the data that a subscriber wants to discard. Something like `data.Discard()` and then, in `EmitAndRelease`, we stop sending that data element to the rest of subscribers. WDYT?

cc @mauriciovasquezbernal

Reply:
Yeah, that error there is just a dummy - I discussed that already with @mauriciovasquezbernal and we came to the conclusion to use a specific error from the datasource package for it (datasource.ErrSkip or something like it). So this would make it skip that event without returning a real error message.

For arrays the same could happen in the `EmitAndRelease()` in that that specific error would shrink the array.